File: ios/Runner/MeshSdkManager.swift
import Flutter
import UIKit
import cnmesh

public class MeshSdkManager: NSObject, FlutterPlugin {
    private var meshSdkManager2: MeshSdkManager2!
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "com.example.ble_scanner/ble", binaryMessenger: registrar.messenger())
        let instance = MeshSdkManager()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    override init() {
        super.init()
        CNMeshSDKKt.initializeMeshSDK()
        meshSdkManager2 = MeshSdkManager2()
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "updateMeshUserNameAndPassword":
            if let args = call.arguments as? [String: Any],
               let username = args["username"] as? String,
               let password = args["password"] as? String {
                meshSdkManager2.updateMeshUserNameAndPassword(username: username, password: password)
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Username and password are required", details: nil))
            }
        case "registerConnectionEvent":
            meshSdkManager2.registerConnectionEvent { meshState in
                DispatchQueue.main.async {
                    switch meshState {
                    case MeshState.Connected():
                        self.showToast(message: "Mesh Connected")
                    case MeshState.Connecting():
                        self.showToast(message: "Mesh Connecting")
                    case MeshState.Disconnected():
                        self.showToast(message: "Mesh Disconnected")
                    default:
                        break
                    }
                }
            }
            result(nil)
        case "startProvisionMode":
            meshSdkManager2.startProvisionMode()
            result(nil)
        case "stopProvisionMode":
            meshSdkManager2.stopProvisionMode()
            result(nil)
        case "getFactoryMeshDevices":
            meshSdkManager2.getFactoryMeshDevices { devices in
                result(devices.map { $0.toDict() })
            }
        case "getProvisionedMeshDevices":
            meshSdkManager2.getProvisionedMeshDevices { devices in
                result(devices.map { $0.toDict() })
            }
        case "getMeshDevices":
            meshSdkManager2.getMeshDevices { devices in
                result(devices.map { $0.toDict() })
            }
        case "getFactoryMeshDevicesWithFilter":
            if let args = call.arguments as? [String: Any],
               let filter = args["filter"] as? [String: Any] {
                let deviceFilter = self.createDeviceFilter(filter)
                meshSdkManager2.getFactoryMeshDevices(deviceFilter: deviceFilter) { devices in
                    result(devices.map { $0.toDict() })
                }
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid filter", details: nil))
            }
        case "provisionDevice":
            if let args = call.arguments as? [String: Any],
               let macAddress = args["macAddress"] as? String,
               let meshAddress = args["meshAddress"] as? Int32,
               let groupAddress = args["groupAddress"] as? Int32 {
                meshSdkManager2.provisionDevice(
                    macAddress: macAddress,
                    meshAddress: meshAddress,
                    provisionCommands: [
                        Commands.AddToGroup(groupAddress: groupAddress, commandDelay: 0)
                    ]
                ) { response in
                    if response.data != nil {
                        result(true)
                    } else {
                        result(FlutterError(code: "PROVISION_FAILED", message: response.message, details: nil))
                    }
                }
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments for provisionDevice", details: nil))
            }
        case "unProvisionDevice":
            if let args = call.arguments as? [String: Any],
               let meshAddress = args["meshAddress"] as? Int32 {
                meshSdkManager2.unProvisionDevice(meshAddress: meshAddress) { response in
                    if response.data != nil {
                        result(true)
                    } else {
                        result(FlutterError(code: "UNPROVISION_FAILED", message: response.message, details: nil))
                    }
                }
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Mesh address is required", details: nil))
            }
        case "sendCommand":
            if let args = call.arguments as? [String: Any],
               let meshAddress = args["meshAddress"] as? Int32,
               let commandType = args["commandType"] as? String,
               let commandParams = args["commandParams"] as? [String: Any] {
                let command = self.createCommand(type: commandType, params: commandParams)
                meshSdkManager2.sendCommand(
                    meshAddress: meshAddress,
                    commands: [command]
                ) { response in
                    if response.data != nil {
                        result(true)
                    } else {
                        result(FlutterError(code: "COMMAND_FAILED", message: response.message, details: nil))
                    }
                }
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments for sendCommand", details: nil))
            }
        case "connectAndSendCommand":
            if let args = call.arguments as? [String: Any],
               let macAddress = args["macAddress"] as? String,
               let commands = args["commands"] as? [[String: Any]] {
                let commandsList = commands.map { self.createCommand(type: $0["type"] as! String, params: $0["params"] as! [String: Any]) }
                meshSdkManager2.connectAndSendCommand(
                    macAddress: macAddress,
                    commands: commandsList
                ) { response in
                    if response.data != nil {
                        result(true)
                    } else {
                        result(FlutterError(code: "COMMAND_FAILED", message: response.message, details: nil))
                    }
                }
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments for connectAndSendCommand", details: nil))
            }
        case "getGateways":
            meshSdkManager2.getGateways { gateways in
                result(gateways.map { $0.toDict() })
            }
        case "getGatewaysWithFilter":
            if let args = call.arguments as? [String: Any],
               let filter = args["filter"] as? [String: Any] {
                let deviceFilter = self.createDeviceFilter(filter)
                meshSdkManager2.getGateways(deviceFilter: deviceFilter) { gateways in
                    result(gateways.map { $0.toDict() })
                }
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid filter", details: nil))
            }
        case "provisionGateway":
            if let args = call.arguments as? [String: Any],
               let macAddress = args["macAddress"] as? String {
                meshSdkManager2.provisionGateway(meshDevice: MeshDevice(macAddress: macAddress)) { response in
                    if response.data != nil {
                        result(true)
                    } else {
                        result(FlutterError(code: "PROVISION_GATEWAY_FAILED", message: response.message, details: nil))
                    }
                }
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Mac address is required", details: nil))
            }
        case "unProvisionGateway":
            if let args = call.arguments as? [String: Any],
               let macAddress = args["macAddress"] as? String {
                meshSdkManager2.unProvisionGateway(macAddress: macAddress) { response in
                    if response.data != nil {
                        result(true)
                    } else {
                        result(FlutterError(code: "UNPROVISION_GATEWAY_FAILED", message: response.message, details: nil))
                    }
                }
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Mac address is required", details: nil))
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func showToast(message: String) {
        DispatchQueue.main.async {
            let alert = UIAlertController(title: nil, message: message, preferredStyle: .alert)
            alert.view.backgroundColor = .black
            alert.view.alpha = 0.5
            alert.view.layer.cornerRadius = 15
            
            if let window = UIApplication.shared.windows.first(where: { $0.isKeyWindow }) {
                window.rootViewController?.present(alert, animated: true)
            }
            
            DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
                alert.dismiss(animated: true)
            }
        }
    }
    
    private func createDeviceFilter(_ filter: [String: Any]) -> DeviceFilter {
        return DeviceFilter(
            productId: filter["productId"] as? String,
            macAddress: filter["macAddress"] as? String,
            rssi: filter["rssi"] as? Int32
        )
    }
    
    private func createCommand(type: String, params: [String: Any]) -> Commands {
        switch type {
        case "Blink":
            return Commands.Blink(commandDelay: Int64(params["delay"] as? Int ?? 0))
        case "On":
            return Commands.On(commandDelay: Int64(params["delay"] as? Int ?? 0))
        case "Off":
            return Commands.Off(commandDelay: Int64(params["delay"] as? Int ?? 0))
        case "Brightness":
            return Commands.Brightness(brightness: Int32(params["brightness"] as? Int ?? 0), commandDelay: Int64(params["delay"] as? Int ?? 0))
        // Add other command types as needed
        default:
            fatalError("Unknown command type: \(type)")
        }
    }
}

extension MeshDevice {
    func toDict() -> [String: Any] {
        return [
            "macAddress": macAddress,
            // Add other properties as needed
        ]
    }
}

File: android/app/src/main/kotlin/com/example/ble_testing/MeshSdkManager.kt
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.widget.Toast
import com.cosmicnode.cnmesh.MeshSdkManager2
import com.cosmicnode.cnmesh.datastore.mesh.manager.MeshState
import com.cosmicnode.cnmesh.domain.Commands
import com.cosmicnode.cnmesh.domain.DeviceFilter
import com.cosmicnode.cnmesh.domain.MeshDevice
import com.cosmicnode.cnmesh.initializeMeshSDK
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class MeshSdkManager(private val context: Context) : MethodCallHandler {
    private var meshSdkManager2: MeshSdkManager2
    private val handler = Handler(Looper.getMainLooper())

    init {
        initializeMeshSDK()
        meshSdkManager2 = MeshSdkManager2()
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "updateMeshUserNameAndPassword" -> {
                val username = call.argument<String>("username")
                val password = call.argument<String>("password")
                if (username != null && password != null) {
                    meshSdkManager2.updateMeshUserNameAndPassword(username, password)
                    result.success(null)
                } else {
                    result.error("INVALID_ARGUMENTS", "Username and password are required", null)
                }
            }
            "registerConnectionEvent" -> {
                meshSdkManager2.registerConnectionEvent { meshState ->
                    handler.post {
                        when (meshState) {
                            MeshState.Connected -> showToast("Mesh Connected")
                            MeshState.Connecting -> showToast("Mesh Connecting")
                            MeshState.Disconnected -> showToast("Mesh Disconnected")
                        }
                    }
                }
                result.success(null)
            }
            "startProvisionMode" -> {
                meshSdkManager2.startProvisionMode()
                result.success(null)
            }
            "stopProvisionMode" -> {
                meshSdkManager2.stopProvisionMode()
                result.success(null)
            }
            "getFactoryMeshDevices" -> {
                meshSdkManager2.getFactoryMeshDevices { devices ->
                    result.success(devices.map { it.toMap() })
                }
            }
            "getProvisionedMeshDevices" -> {
                meshSdkManager2.getProvisionedMeshDevices { devices ->
                    result.success(devices.map { it.toMap() })
                }
            }
            "getMeshDevices" -> {
                meshSdkManager2.getMeshDevices { devices ->
                    result.success(devices.map { it.toMap() })
                }
            }
            "getFactoryMeshDevicesWithFilter" -> {
                val filter = call.argument<Map<String, Any>>("filter")
                val deviceFilter = createDeviceFilter(filter)
                meshSdkManager2.getFactoryMeshDevices(deviceFilter) { devices ->
                    result.success(devices.map { it.toMap() })
                }
            }
            "provisionDevice" -> {
                val macAddress = call.argument<String>("macAddress")
                val meshAddress = call.argument<Int>("meshAddress")
                val groupAddress = call.argument<Int>("groupAddress")
                if (macAddress != null && meshAddress != null && groupAddress != null) {
                    meshSdkManager2.provisionDevice(
                        MeshDevice(macAddress),
                        meshAddress,
                        listOf(Commands.AddToGroup(groupAddress))
                    ) { response ->
                        if (response.data != null) {
                            result.success(true)
                        } else {
                            result.error("PROVISION_FAILED", response.message?.toString(), null)
                        }
                    }
                } else {
                    result.error("INVALID_ARGUMENTS", "Invalid arguments for provisionDevice", null)
                }
            }
            "unProvisionDevice" -> {
                val meshAddress = call.argument<Int>("meshAddress")
                if (meshAddress != null) {
                    meshSdkManager2.unProvisionDevice(meshAddress) { response ->
                        if (response.data != null) {
                            result.success(true)
                        } else {
                            result.error("UNPROVISIONED_FAILED", response.message?.toString(), null)
                        }
                    }
                } else {
                    result.error("INVALID_ARGUMENTS", "Mesh address is required", null)
                }
            }
            "sendCommand" -> {
                val meshAddress = call.argument<Int>("meshAddress")
                val commandType = call.argument<String>("commandType")
                val commandParams = call.argument<Map<String, Any>>("commandParams")
                if (meshAddress != null && commandType != null) {
                    val command = createCommand(commandType, commandParams)
                    meshSdkManager2.sendCommand(meshAddress, listOf(command)) { response ->
                        if (response.data != null) {
                            result.success(true)
                        } else {
                            result.error("COMMAND_FAILED", response.message?.toString(), null)
                        }
                    }
                } else {
                    result.error("INVALID_ARGUMENTS", "Invalid arguments for sendCommand", null)
                }
            }
            "connectAndSendCommand" -> {
                val macAddress = call.argument<String>("macAddress")
                val commands = call.argument<List<Map<String, Any>>>("commands")
                if (macAddress != null && commands != null) {
                    val commandsList = commands.map { createCommand(it["type"] as String, it["params"] as Map<String, Any>) }
                    meshSdkManager2.connectAndSendCommand(macAddress, commandsList) { response ->
                        if (response.data != null) {
                            result.success(true)
                        } else {
                            result.error("COMMAND_FAILED", response.message?.toString(), null)
                        }
                    }
                } else {
                    result.error("INVALID_ARGUMENTS", "Invalid arguments for connectAndSendCommand", null)
                }
            }
            "getGateways" -> {
                meshSdkManager2.getGateways { gateways ->
                    result.success(gateways.map { it.toMap() })
                }
            }
            "getGatewaysWithFilter" -> {
                val filter = call.argument<Map<String, Any>>("filter")
                val deviceFilter = createDeviceFilter(filter)
                meshSdkManager2.getGateways(deviceFilter) { gateways ->
                    result.success(gateways.map { it.toMap() })
                }
            }
            "provisionGateway" -> {
                val macAddress = call.argument<String>("macAddress")
                if (macAddress != null) {
                    meshSdkManager2.provisionGateway(MeshDevice(macAddress)) { response ->
                        if (response.data != null) {
                            result.success(true)
                        } else {
                            result.error("PROVISION_GATEWAY_FAILED",
                                response.message?.toString(), null)
                        }
                    }
                } else {
                    result.error("INVALID_ARGUMENTS", "Mac address is required", null)
                }
            }
            "unProvisionGateway" -> {
                val macAddress = call.argument<String>("macAddress")
                if (macAddress != null) {
                    meshSdkManager2.unProvisionGateway(macAddress) { response ->
                        if (response.data != null) {
                            result.success(true)
                        } else {
                            result.error("UNPROVISIONED_GATEWAY_FAILED",
                                response.message?.toString(), null)
                        }
                    }
                } else {
                    result.error("INVALID_ARGUMENTS", "Mac address is required", null)
                }
            }
            else -> result.notImplemented()
        }
    }

    private fun showToast(message: String) {
        Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
    }

    private fun MeshDevice.toMap(): Map<String, String?> {
        return mapOf(
            "macAddress" to macAddress,
            // Add other properties as needed
        )
    }

    private fun createDeviceFilter(filter: Map<String, Any>?): DeviceFilter {
        return DeviceFilter(
            productId = (filter?.get("productId") as? List<*>)?.mapNotNull { it as? Int },
            macAddress = filter?.get("macAddress") as? String,
            rssi = filter?.get("rssi") as? Int
        )
    }

    private fun createCommand(type: String, params: Map<String, Any>?): Commands {
        return when (type) {
            "Blink" -> Commands.Blink(params?.get("delay") as? Long ?: 0L)
            "On" -> Commands.On(params?.get("delay") as? Long ?: 0L)
            "Off" -> Commands.Off(params?.get("delay") as? Long ?: 0L)
            "Brightness" -> Commands.Brightness(params?.get("brightness") as? Int ?: 0, params?.get("delay") as? Long ?: 0L)
            // Add other command types as needed
            else -> throw IllegalArgumentException("Unknown command type: $type")
        }
    }
}